### ISSUE TITLE
Fail to find call media with videoIncomingWindowId in Call.onCallMediaState on Android

### ISSUE THREAD

**Issue by ArtsemKurantsou** [pjsip/pjproject#2529]
**Describe the bug**
I'm working on the Android app that uses PjSip for calls and got a bug report from one of the user that he sometimes can't see the incoming video from an incoming call. I've added additional logs to the native part and figured out that there is an issue with VideoWindow initialization on the Java side. In the logs, I can see that the video window is successfully created and initialized in the native part, but my app can't find it on the Java side. I tracked down to the following state:

1. pjsua_call_on_media_update is called once SDP negotiation is completed
2. pjsua_vid_channel_update is called
3. create_vid_win is also called and a native video window is created (in the logs the message is "stream window id 0 created for cap_dev=-1 rend_dev=0")
4. Call.onCallMediaState (Java side) is called (in the logs the message is "PjSipCall	onCallMediaState: org.pjsip.pjsua2.OnCallMediaStateParam@61b188b"). In this callback, I call Call.getInfo to get call medias and enumerate through it to find one with type - pjmedia_type.PJMEDIA_TYPE_VIDEO and videoIncomingWindowId() != pjsua2.INVALID_ID. One I found one I do log another message with the status of this media but for this "failed" call I don't see this message. So it means that this media is absent, has wrong type or videoncomingWindowId is invalid.

I tried to find anything in the native code myself self but I don't see anything that can go wrong there. 

**To Reproduce**
I wasn't able to reproduce this but the user is able to do it something like 2 out of 10 calls.  

**Expected behavior**
Call media with type pjmedia_type.PJMEDIA_TYPE_VIDEO and valid videoIncomingWindowId() can be found in the onCallMediaState callback.

**Logs/Screenshots**
Here is a log file for such a call.
[logs_for_video_window_initialization_bug.txt](https://github.com/pjsip/pjproject/files/5262330/logs_for_video_window_initialization_bug.txt)
In this logs there additional messages like "trying to render frame", "Skipping frame. Stream no display." and "decode result" that was added by me to find if there is an issue with the video itself. 

**Smartphone**

- Android 10 (SDK 29)
- PJSIP
  - version: 2.9
  - `configure` script params: TARGET_ABI=$arch APP_PLATFORM=android-$level ./configure-android --use-ndk-cflags --with-opus=$opus_path --with-openh264=$h264_path --with-ssl=$open_ssl_path
  - `config_site.h` contents: 
```C
#define PJ_CONFIG_ANDROID 1
#define PJMEDIA_HAS_VIDEO 1
#define PJ_HAS_IPV6 1
#define PJMEDIA_CODEC_OPUS_DEFAULT_BIT_RATE 64000
#define PJMEDIA_RTP_PT_TELEPHONE_EVENTS 121
#define SIP_USE_ICE 1
#include <pj/config_site_sample.h>
```
  - related third party libraries & versions: OpenSSL 1.1.1g, OpenH264 1.6.0, Opus 1.2.1

- Device: Samsung SM-G975U

**Additional context**
Android NDK 13b was used for the build

**Comment by ArtsemKurantsou** [comment]
I've added some logs to the native part of PjSip to log call medias in pjsua_call_get_info and on the Java side in Call.onCallMediaState and find out that on the native part both active and provisional medias has 2 elements (audio and video) but on Java side only audio one is present. So it means that video one got lost in between. I think it can have something to do with conversion code (maybe something with vector usage).

**Comment by sauwming** [comment]
Where exactly in the native code do you add the log?

The conversion to pjsua2 object happens here:
https://github.com/pjsip/pjproject/blob/master/pjsip/src/pjsua2/call.cpp#L299
So, check if `pci.media_cnt` and `pci.prov_media_cnt` are 2.

Then the conversion to Java object happens here:
`pjsip-apps/src/swig/java/android/app/src/main/java/org/pjsip/pjsua2/CallInfo.java`
```
  public CallMediaInfoVector getMedia() {
    long cPtr = pjsua2JNI.CallInfo_media_get(swigCPtr, this);
    return (cPtr == 0) ? null : new CallMediaInfoVector(cPtr, false);
  }
```
So if `CallMediaInfoVector(cptr)` only has one element, then something is really wrong here.

- Can you tell us the SWIG version you're using? Try using the latest SWIG version if possible.
- Make sure that the library is built cleanly from scratch. There may be a mix between the old and new build, or there can accidentally be two installations. So cleanup all the previous installation and files as well.

**Comment by ArtsemKurantsou** [comment]
Hi @sauwming. I used SWIG 3.0.12 for the build and for the build process I use my own script that always copies sources to the separate directory for the build so build happens only on "clean" sources. 
I used the following code for logging in `pjsua_call_get_info`
```C
    /* Build array of active media info */
    info->media_cnt = 0;
	PJ_LOG(5, (THIS_FILE, "pjsua_call_get_info: active media count - %d", call->med_cnt));
    for (mi=0; mi < call->med_cnt &&
	       info->media_cnt < PJ_ARRAY_SIZE(info->media); ++mi)
    {
	pjsua_call_media *call_med = &call->media[mi];

	PJ_LOG(5, (THIS_FILE, "pjsua_call_get_info: active media %d, type - %d, status - %d, render win id - %d", mi, call_med->type, call_med->state, call_med->strm.v.rdr_win_id));
       
       //...
    }
   
    //...

    /* Build array of provisional media info */
    info->prov_media_cnt = 0;
	PJ_LOG(5, (THIS_FILE, "pjsua_call_get_info: provisional media count - %d", call->med_prov_cnt));
    for (mi=0; mi < call->med_prov_cnt &&
	       info->prov_media_cnt < PJ_ARRAY_SIZE(info->prov_media); ++mi)
    {
	pjsua_call_media *call_med = &call->media_prov[mi];

	PJ_LOG(5, (THIS_FILE, "pjsua_call_get_info: provisional media %d, type - %d, status - %d, render win id - %d", mi, call_med->type, call_med->state, call_med->strm.v.rdr_win_id));
        //...
    }
```
On the Java side in `Call.onCallMediaState` I added the following code
```Java
    @Override
    public void onCallMediaState(OnCallMediaStateParam prm) {
        //...
        for (int i = 0; i < cmiv.size(); i++) {
            CallMediaInfo cmi = cmiv.get(i);
            Log.d(TAG, String.format("onCallMediaState: media - %d, type - %d, status - %d, incomingWindowId - %d",
                    i, cmi.getType(), cmi.getStatus(), cmi.getVideoIncomingWindowId()));
           //...
        }
    }
```
In the logs that I've got from the user, I see the following:
```
17:46:18:556	DEBUG	PjSipLogWriter	 UTC=1600908378555 17:46:18.555           pjsua_call.c !Answering call 3: code=200	
//...
17:46:18:557	DEBUG	PjSipLogWriter	 UTC=1600908378557 17:46:18.557            pjsua_vid.c  ....Video channel update..	
17:46:18:557	DEBUG	PjSipLogWriter	 UTC=1600908378557 17:46:18.557           openh264.cpp  .....Opening codec..	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560                  rtp.c  .....pjmedia_rtp_session_init: ses=0x78f1d26d3c, default_pt=103, ssrc=0x9f015c0	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560     vstdec0x791bbb9828  .....Decoding channel created 1456x1456 I420<-H264 22/1(~22)fps	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560     vstenc0x791bbb9828  .....Encoding channel created 352x288 I420->H264 15/1(~15)fps	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560        udp0x793456f800  .....SO_SNDBUF set to 229376	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560        udp0x793456f800  .....SO_RCVBUF set to 229376	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378559 17:46:18.559                  rtp.c  .....pjmedia_rtp_session_init: ses=0x791bbb9f3c, default_pt=103, ssrc=0x9f015c0	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560           vid_stream.c  .....Video stream vstrm0x791bbb9828 created	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560     vstenc0x791bbb9828  .....Encoder stream started	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560            pjsua_vid.c  .....Setting up RX..	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560     vstdec0x791bbb9828  .....Decoder stream started	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560            pjsua_vid.c  ......Creating video window: type=stream, cap_id=-1, rend_id=0	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560             vid_port.c  .......Opening device OpenGL renderer [OpenGL] for render: format=I420, size=1456x1456 @22:1 fps	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560     android_opengl.cpp  .......Re-initializing OpenGL due to format change	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560     android_opengl.cpp  .......Android OpenGL ES renderer successfully created	
17:46:18:560	DEBUG	PjSipLogWriter	 UTC=1600908378560 17:46:18.560             vid_port.c  .......Device OpenGL renderer [OpenGL] opened: format=BGRA, size=1456x1456 @22:1 fps	
17:46:18:563	DEBUG	PjSipLogWriter	 UTC=1600908378563 17:46:18.562             vid_conf.c  .......Added port 0 (OpenGL renderer)	
17:46:18:563	DEBUG	PjSipLogWriter	 UTC=1600908378563 17:46:18.563            pjsua_vid.c  .......Window 0 created	
17:46:18:563	DEBUG	PjSipLogWriter	 UTC=1600908378563 17:46:18.563            pjsua_vid.c  .......stream window id 0 created for cap_dev=-1 rend_dev=0	
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564             vid_conf.c  ......Added port 1 (vstdec0x791bbb9828)	
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564             vid_conf.c  ......Updating render state for port id 0 (1 sources)..	
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564             vid_conf.c  ......Port 1 (vstdec0x791bbb9828) transmitting to port 0 (OpenGL renderer)	
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564     android_opengl.cpp  ......Starting Android opengl stream	
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564             vid_conf.c  ......This port only has single source with matched format & size, no conversion needed	
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564            pjsua_vid.c  .....Setting up TX..	
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564             vid_conf.c  ......Added port 2 (vstenc0x791bbb9828)	
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564     vstenc0x791bbb9828  .....Encoder stream paused	
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564          pjsua_media.c  ....Video updated, stream #1: H264 (sendrecv)
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564           pjsua_call.c  ...pjsua_call_get_info: active media count - 2	
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564           pjsua_call.c  ...pjsua_call_get_info: active media 0, type - 1, status - 1, render win id - -1	
17:46:18:564	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564           pjsua_call.c  ...pjsua_call_get_info: active media 1, type - 2, status - 1, render win id - 0	
17:46:18:565	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564           pjsua_call.c  ...pjsua_call_get_info: provisional media count - 2	
17:46:18:565	DEBUG	PjSipLogWriter	 UTC=1600908378564 17:46:18.564           pjsua_call.c  ...pjsua_call_get_info: provisional media 0, type - 1, status - 1, render win id - -1	
17:46:18:565	DEBUG	PjSipLogWriter	 UTC=1600908378565 17:46:18.565           pjsua_call.c  ...pjsua_call_get_info: provisional media 1, type - 2, status - 1, render win id - 0	
17:46:18:565	DEBUG	PjSipCall	onCallMediaState: media - 0, type - PJMEDIA_TYPE_AUDIO, status - PJSUA_CALL_MEDIA_ACTIVE, incomingWindowId - -1	
```

**Comment by FredWH** [comment]
![cmiv2](https://user-images.githubusercontent.com/85405161/134830215-e13d60ea-e6b8-4deb-876b-eb797d3f3257.png)
There is a similar no video issue, but it happen in several android phones, like Samsung S10, and about 5 times in 1000 calls.
When the issue happened, we found cmiv.size() was changed from 2 to 0, during loop.
There is a workaround: getCaptureDevMedia().startTransmit(...) need about 40ms, move it out of loop, new VideoWindow before cmiv.size() changed to 0.
